### PR TITLE
support token based login using format Bearer:<token> in password field

### PR DESCRIPTION
**Pull request by ashiquzzaman33**

### CODE REVIEW

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 1]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
+      logger.info(String
+          .format("Invalid login. namespace: %s, username: %s", currentAuth.getTeamId(), username));
+      throw new UsernameNotFoundException("Username or password didn't match");
+    }
+    return username;
+  }
+
+  public boolean isToken(String password) {
+    if (password.startsWith(BearerPrefix)
+        && password.substring(BearerPrefix.length()).length() == 32) {
```
Comment (sadlil) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: Use Const `BearerTokenLength`

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 2]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
+    }
+
+    ConduitGrpc.ConduitBlockingStub conduitBlockingStub = ConduitGrpc.newBlockingStub(channel).
+        withCallCredentials(
+            MoreCallCredentials.from(new TokenCredential(currentAuth.getTeamId(), token)));
+    ConduitWhoAmIResponse whoAmIResponse = conduitBlockingStub
+        .whoAmI(VoidRequest.newBuilder().build());
+    if (!whoAmIResponse.getStatus().getCode().equals("0")) {
+      logger.info(String
+          .format("error while calling WhoAmI. Namespace: %s,  status message: %s",
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: e -> `E`

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 3]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
+
+  private String tokenLogin(String token) throws UsernameNotFoundException {
+    TokenRequest tokenRequest = TokenRequest.newBuilder()
+        .setNamespace(currentAuth.getTeamId())
+        .setToken(token).build();
+    Status tokenStatus = authenticationBlockingStub.token(tokenRequest).getStatus();
+    if (!tokenStatus.getCode().equals("0")) {
+      logger.info(String
+          .format("Invalid token. namespace: %s", currentAuth.getTeamId()));
+      throw new UsernameNotFoundException("invalid token");
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: i -> `I`

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 4]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
       builder.negotiationType(NegotiationType.PLAINTEXT);
     }
     channel = builder.build();
+    authenticationBlockingStub = AuthenticationGrpc.newBlockingStub(channel);
+  }
+
+  private String tokenLogin(String token) throws UsernameNotFoundException {
+    TokenRequest tokenRequest = TokenRequest.newBuilder()
+        .setNamespace(currentAuth.getTeamId())
+        .setToken(token).build();
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: .build() should go to the next line.

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 5]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
-              .negotiationType(NegotiationType.TLS);
+          .negotiationType(NegotiationType.TLS);
     } else {
       builder.negotiationType(NegotiationType.PLAINTEXT);
     }
     channel = builder.build();
+    authenticationBlockingStub = AuthenticationGrpc.newBlockingStub(channel);
+  }
+
+  private String tokenLogin(String token) throws UsernameNotFoundException {
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: tokenLogin -> checkToken()

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 6]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
+    if (!whoAmIResponse.getStatus().getCode().equals("0")) {
+      logger.info(String
+          .format("error while calling WhoAmI. Namespace: %s,  status message: %s",
+              currentAuth.getTeamId(), whoAmIResponse.getStatus().getMessage()));
+      throw new UsernameNotFoundException("error: ", whoAmIResponse.getStatus().getMessage());
+    }
+    return whoAmIResponse.getUser().getUserName();
+  }
+
+  private String basicLogin(String username, String password) throws UsernameNotFoundException {
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: basicLogin() -> `checkPassword()`

File: src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java [hunk 7]
```diff
@@ -104,11 +101,61 @@ private void initializeAuthAndChannel() throws URISyntaxException, SSLException
...
+    LoginResponse response = authenticationBlockingStub.login(request);
+    if (!response.getStatus().getCode().equals("0")) {
+      logger.info(String
+          .format("Invalid login. namespace: %s, username: %s", currentAuth.getTeamId(), username));
+      throw new UsernameNotFoundException("Username or password didn't match");
+    }
+    return username;
+  }
+
+  public boolean isToken(String password) {
```
Comment (TamalSaha) [inline src/main/java/com/appscode/ci/plugins/authn/AppsCodeSecurityRealm.java]: `return password.startsWith(BearerPrefix) && password.substring(BearerPrefix.length()).length() == 32`

**Pull request merged**